examples/timer_functional/src/main.rs

```diff
@@ -0,0 +1,166 @@
+}
+
+enum TimerAction {
+    Add(&'static str),
+    Cancel,
+    SetInterval(Interval),
+    SetTimeout(Timeout),
+    TimeoutDone,
+}
+
+#[derive(Clone, Debug)]
+struct TimerState {
+    messages: Vec<&'static str>,
+    interval_handle: Option<Rc<Interval>>,
+    timeout_handle: Option<Rc<Timeout>>,
+}
+
+impl PartialEq for TimerState {
+    fn eq(&self, other: &Self) -> bool {
+        self.messages == other.messages
+            && self.interval_handle.is_some() == other.interval_handle.is_some()
+    }
+}
+
+impl Reducible for TimerState {
+    type Action = TimerAction;
+
+    fn reduce(self: Rc<Self>, action: TimerAction) -> Rc<Self> {
+        match action {
+            TimerAction::Add(message) => {
+                let mut messages = self.messages.clone();
+                messages.push(message);
+                Rc::new(TimerState {
+                    messages,
+                    interval_handle: self.interval_handle.clone(),
+                    timeout_handle: self.timeout_handle.clone(),
+                })
+            }
+            TimerAction::SetInterval(t) => Rc::new(TimerState {
+                messages: vec!["Interval started!"],
+                interval_handle: Some(Rc::from(t)),
+                timeout_handle: self.timeout_handle.clone(),
+            }),
+            TimerAction::SetTimeout(t) => Rc::new(TimerState {
+                messages: vec!["Timer started!!"],
+                interval_handle: self.interval_handle.clone(),
+                timeout_handle: Some(Rc::from(t)),
+            }),
+            TimerAction::TimeoutDone => {
+                let mut messages = self.messages.clone();
+                messages.push("Done!");
+                Rc::new(TimerState {
+                    messages,
+                    interval_handle: self.interval_handle.clone(),
+                    timeout_handle: None,
+                })
+            }
+            TimerAction::Cancel => {
+                let mut messages = self.messages.clone();
+                messages.push("Canceled!");
+                Rc::new(TimerState {
+                    messages,
+                    interval_handle: None,
+                    timeout_handle: None,
+                })
+            }
+        }
+    }
+}
+
+#[function_component(Clock)]
+fn clock() -> Html {
+    let time = use_state(get_current_time);
+
+    {
+        let time = time.clone();
+        use_effect_with_deps(
+            |_| {
+                Interval::new(1000, move || time.set(get_current_time())).forget();
+            },
+            (),
+        );
+    }
+    html!(
+        <div id="time">{ time.as_str() }</div>
+    )
+}
+
+#[function_component]
+fn App() -> Html {
+    let state = use_reducer(|| TimerState {
+        messages: Vec::new(),
+        interval_handle: None,
+        timeout_handle: None,
+    });
+
+    let messages: Html = state
+        .messages
+        .iter()
+        .map(|message| html! { <p>{ message }</p> })
```

Comment:
When `Html` is created from an iterator, it should be assigned a key for each element. Missing keys will cause all elements to be updated when an element is inserted at the end of the array.